### ISSUE TITLE
Find project

### ISSUE THREAD

**Issue by huseingharibshah** [huseingharibshah/Parameters#2]
Git hub account id:::

XG boost hearr disease
https://github.com/diego-taquiri/HeartDiseaseMLInterpretation/commits?author=diego-taquiri

Ensemble learning on loan dataset
https://github.com/madhur02/ensemble-learning-loan-dataset/commits?author=madhur02

numpy
pandas
scikit-learn
xgboost
joblib
matplotlib
seaborn
kaggle   # اگر می‌خواهی از Kaggle API دیتاست بگیری

# train_xgboost.py
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import accuracy_score, classification_report
import xgboost as xgb
import joblib
import argparse

parser = argparse.ArgumentParser()
parser.add_argument("--data", type=str, default="data/dataset.csv")
parser.add_argument("--output", type=str, default="models/xgb_model.joblib")
args = parser.parse_args()

df = pd.read_csv(args.data)
# فرض: ستون هدف نامش 'target' است — در صورت فرق، تغییر بده
X = df.drop(columns=["target"])
y = df["target"]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

xgb_clf = xgb.XGBClassifier(use_label_encoder=False, eval_metric="logloss", random_state=42)

# مثالِ جستجوی هایپرپارامتر ساده (GridSearch)
param_grid = {
    "n_estimators": [50, 100],
    "max_depth": [3, 6],
    "learning_rate": [0.01, 0.1]
}

grid = GridSearchCV(xgb_clf, param_grid, cv=3, scoring="accuracy", n_jobs=2, verbose=1)
grid.fit(X_train, y_train)

best = grid.best_estimator_
y_pred = best.predict(X_test)

print("Best params:", grid.best_params_)
print("Test accuracy:", accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))

# ذخیره مدل
joblib.dump(best, args.output)
print("Model saved to", args.output)

# train_ensemble.py
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier, StackingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report
import joblib
import argparse

parser = argparse.ArgumentParser()
parser.add_argument("--data", type=str, default="data/dataset.csv")
parser.add_argument("--output", type=str, default="models/stack_model.joblib")
args = parser.parse_args()

df = pd.read_csv(args.data)
X = df.drop(columns=["target"])
y = df["target"]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

estimators = [
    ("rf", RandomForestClassifier(n_estimators=100, random_state=42)),
    ("gb", GradientBoostingClassifier(n_estimators=100, random_state=42))
]

stack = StackingClassifier(estimators=estimators, final_estimator=LogisticRegression(), n_jobs=2)
stack.fit(X_train, y_train)
y_pred = stack.predict(X_test)

print("Test accuracy:", accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))
joblib.dump(stack, args.output)
print("Saved stacking model to", args.output)

name: Train models

on:
  push:
    branches: [ main ]
  workflow_dispatch:

jobs:
  setup-and-train:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4

      - name: Set up Python
        uses: actions/setup-python@v4
        with:
          python-version: "3.10"

      - name: Install dependencies
        run: |
          python -m pip install --upgrade pip
          pip install -r requirements.txt

      # اگر می‌خواهی دیتاست را از Kaggle بگیری (اختیاری)
      - name: Download dataset from Kaggle (optional)
        if: ${{ secrets.KAGGLE_USERNAME && secrets.KAGGLE_KEY }}
        env:
          KAGGLE_USERNAME: ${{ secrets.KAGGLE_USERNAME }}
          KAGGLE_KEY: ${{ secrets.KAGGLE_KEY }}
        run: |
          mkdir -p ~/.kaggle
          echo "{\"username\":\"$KAGGLE_USERNAME\",\"key\":\"$KAGGLE_KEY\"}" > ~/.kaggle/kaggle.json
          chmod 600 ~/.kaggle/kaggle.json
          # مثال: kaggle datasets download -d altruistdelhite04/loan-prediction-problem-dataset -p data --unzip
          # دستور بالا را بر اساس dataset مورد نظر تغییر بده

      - name: Create model dir
        run: mkdir -p models

      - name: Train XGBoost
        run: |
          python train_xgboost.py --data data/dataset.csv --output models/xgb_model.joblib

      - name: Train Ensemble
        run: |
          python train_ensemble.py --data data/dataset.csv --output models/stack_model.joblib

      - name: Upload trained models as artifact
        uses: actions/upload-artifact@v4
        with:
          name: trained-models
          path: models/
